3rdparty/p256-m/p256-m_driver_entrypoints.c

```diff
@@ -0,0 +1,238 @@
+        return status;
+    }
+
+    status = p256_to_psa_error(
+        p256_ecdh_shared_secret(shared_secret, key_buffer, peer_key+1));
+    if (status == PSA_SUCCESS) {
+        *shared_secret_length = 32;
+    }
+
+    return status;
+}
+
+psa_status_t p256_transparent_sign_hash(
+    const psa_key_attributes_t *attributes,
+    const uint8_t *key_buffer,
+    size_t key_buffer_size,
+    psa_algorithm_t alg,
+    const uint8_t *hash,
+    size_t hash_length,
+    uint8_t *signature,
+    size_t signature_size,
+    size_t *signature_length)
+{
+    /* We don't use these arguments, but the specification mandates the
+     * sginature of driver entry-points. (void) used to avoid compiler
+     * warning. */
+    (void) attributes;
+    (void) alg;
+
+    psa_status_t status = PSA_ERROR_NOT_SUPPORTED;
+    if (key_buffer_size != 32 || signature_size != 64) {
+        return status;
+    }
+
+    status = p256_to_psa_error(
+        p256_ecdsa_sign(signature, key_buffer, hash, hash_length));
+    if (status == PSA_SUCCESS) {
+        *signature_length = 64;
+    }
+
+    return status;
+}
+
+/*  This function expects the key buffer to contain a 65 byte public key,
+ *  as exported by psa_export_public_key() */
+static psa_status_t p256_verify_hash_with_public_key(
+    const uint8_t *key_buffer,
+    size_t key_buffer_size,
+    const uint8_t *hash,
+    size_t hash_length,
+    const uint8_t *signature,
+    size_t signature_length)
+{
+    psa_status_t status = PSA_ERROR_NOT_SUPPORTED;
+    if (key_buffer_size != 65 || signature_length != 64 || *key_buffer != 0x04) {
+        return status;
+    }
+
+    const uint8_t *public_key_buffer = key_buffer + 1;
+    status = p256_to_psa_error(
+        p256_ecdsa_verify(signature, public_key_buffer, hash, hash_length));
+
+    return status;
+}
+
+psa_status_t p256_transparent_verify_hash(
+    const psa_key_attributes_t *attributes,
+    const uint8_t *key_buffer,
+    size_t key_buffer_size,
+    psa_algorithm_t alg,
+    const uint8_t *hash,
+    size_t hash_length,
+    const uint8_t *signature,
+    size_t signature_length)
+{
+    /* We don't use this argument, but the specification mandates the signature
+     * of driver entry-points. (void) used to avoid compiler warning. */
+    (void) alg;
+
+    psa_status_t status;
+    uint8_t *public_key_buffer = NULL;
+    size_t public_key_buffer_size = 65;
+    public_key_buffer = mbedtls_calloc(1, public_key_buffer_size);
+    if (public_key_buffer == NULL) {
+        return PSA_ERROR_INSUFFICIENT_MEMORY;
+    }
+    size_t *public_key_length = NULL;
+    public_key_length = mbedtls_calloc(1, sizeof(size_t));
+    if (public_key_length == NULL) {
+        return PSA_ERROR_INSUFFICIENT_MEMORY;
+    }
+    *public_key_length = 65;
+
+    /*  The contents of key_buffer may either be the 32 byte private key
+     *  (keypair representation), or the 65 byte public key. To ensure the
+     *  latter is obtained, the public key is exported. */
```

Comment:
You're right, the logic in the core is wrong. https://github.com/Mbed-TLS/mbedtls/issues/8307

Sorry, I missed your comment back then and just happened to read it now. Which is fortunate because this came up again rcecently, I forget where, and I forgot to file an issue at the time. That's fixed now.

Reply:
I read 8307 and either I've misunderstood it, or it's not the same thing as what I was trying to say here.

My point here is that the `"verify_hash"` entry point, which needs a public key, is sometimes passed a keypair instead. Then it needs to somehow convert extract the public key from that before it can call the actual verification function.

If the contract for the `"verify_hash"` entry point is that the code can pass it either a public key or a keypair, then each driver implementing that entry point will have to include some logic to check what it was passed, and extract the public key if necessary. To me, it seems to place extra work on driver writers for no good reason.

I'm suggesting we change the contract so that `"verify_hash"` (and other entry points that naturally expect a public key) would always be passed a public key. If the user calls `psa_verify_hash()` with a keypair object, then it would be the core's job to notice, somehow get the corresponding public and pass the public key to the driver's `"verify_hash"` - then the driver wouldn't have to worry about what it's getting and could simply do its main job of actually performing the verification.

Wdyt?

Reply:
Ah, right, I hadn't looked closely at the context: verify vs export-public.

Filed as https://github.com/Mbed-TLS/mbedtls/issues/8317

Reply:
Thanks!